modules/signoz/main.tf

```diff
@@ -0,0 +1,114 @@
+locals {
+  alertmanager_enabled = var.smtp_from != "" && var.smtp_user != "" && var.smtp_password != ""
+}
+
+resource "kubernetes_namespace" "signoz" {
+  metadata {
+    name = var.namespace
+  }
+}
+
+resource "kubectl_manifest" "signoz-deployment" {
+  depends_on = [kubernetes_namespace.signoz]
+  
+  yaml_body = <<YAML
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: signoz
+  namespace: argocd
+spec:
+  project: default
+  %{if var.auto_deploy}
+  syncPolicy:
+    automated:
+      prune: ${var.auto_prune}
+  %{endif}
+  sources:
+  - repoURL: 'https://charts.signoz.io'
+    chart: signoz
+    targetRevision: 0.50.0
+    helm:
+      releaseName: signoz
+      # Extra parameters to set (same as setting through values.yaml, but these take precedence)
+      parameters:
+      - name: "clickhouse.password"
+        value: ${random_password.clickhouse-admin-password.result}
+      %{if local.alertmanager_enabled}
+      - name: "alertmanager.enabled"
+        value: "true"
+      - name: "alertmanager.additionalEnvs.ALERTMANAGER_SMTP_FROM"
+        value: ${var.smtp_from}
+      - name: "alertmanager.additionalEnvs.ALERTMANAGER_SMTP_AUTH_USERNAME"
+        value: ${var.smtp_user}
+      - name: "alertmanager.additionalEnvs.ALERTMANAGER_SMTP_AUTH_PASSWORD"
+        value: ${var.smtp_password}
+      %{else}
+      - name: "alertmanager.enabled"
+        value: "false"
+      %{endif}
+      valueFiles:
+      - $values/modules/signoz/templates/values.yaml
+  - repoURL: 'https://github.com/Sage-Bionetworks-Workflows/eks-stack.git'
+    targetRevision: ${var.git_revision}
+    ref: values
+  %{if var.enable_otel_ingress}
+  - repoURL: 'https://github.com/Sage-Bionetworks-Workflows/eks-stack.git'
+    targetRevision: ${var.git_revision}
+    path: modules/signoz/resources-otel-ingress
+    kustomize:
+      patches:
+      - target:
+          kind: ReferenceGrant
+        patch: |-
+          - op: replace
+            path: /spec/from/0/namespace
+            value: ${var.gateway_namespace}
+      - target:
+          kind: HTTPRoute
+        patch: |-
+          - op: replace
+            path: /metadata/namespace
+            value: ${var.gateway_namespace}
+          - op: replace
+            path: /spec/rules/0/backendRefs/0/namespace
+            value: ${var.namespace}
+      - target:
+          kind: SecurityPolicy
+        patch: |-
+          - op: replace
+            path: /metadata/namespace
+            value: ${var.gateway_namespace}
+          - op: replace
+            path: /spec/jwt/providers
+            value:
+              - name: auth0
+                remoteJWKS:
+                  uri: ${var.auth0_jwks_uri}
+                audiences:
+                  - ${var.cluster_name}-telemetry

```

Comment:
Adding this security policy let's us do filtering on this specific HTTP route for telemetry incoming to the cluster.

1. When a request enters the cluster this `SecurityPolicy` is evaluated since a request matches to the `http-route.yaml` defined in the `resources-otel-ingress` subdirectory. This allows us to test for one or more specific keys in the incoming JWT. We are verifying that audience is as expected


Testing the export a data with a valid JWT that does not have the expected audience fails with an HTTP 403:
`ERROR:opentelemetry.exporter.otlp.proto.http.trace_exporter:Failed to export batch code: 403, reason: Audiences in Jwt are not allowed`


@zaro0508 This is a key area of the security behind access to submit telemetry data to the cluster, and an area to pay close attention to.

In Auth0 I have an "API": https://auth0.com/docs/get-started/apis

> An API is an entity that represents an external resource, capable of accepting and responding to protected resource requests made by applications. In the [OAuth2 specification](https://tools.ietf.org/html/rfc6749), an API maps to the Resource Server.

For each JWT that is exchanged for this API it has a specific audience that I can filter for here.

**There is an alternative approach to this I was wanting feedback on:**

1. Instead of basing access on the "audience" portion of the JWT claim, I could instead base this on scopes: https://auth0.com/docs/get-started/apis/scopes#ways-to-use-scopes

> In an [API](https://auth0.com/docs/api), to implement access control. In this case, you need to define custom scopes for your API and then identify these scopes so that calling applications can use them.

2. Scopes would allow me to add a scope of `write:telemetry` to the JWT
3. By using scopes instead of audience I could use this as the source of authorization, instead of my current idea of using audience to handle this filtering.

Reply:
This was addressed as the latest version of envoy gateway supports scope based authorization via SecurityPolicies:

https://github.com/Sage-Bionetworks-Workflows/eks-stack/pull/48/files